trio/tests/tools/test_gen_exports.py

```diff
@@ -0,0 +1,233 @@
+import ast
+import astor
+import pytest
+import os
+import sys
+
+from shutil import copyfile
+from trio._tools.gen_exports import (
+    is_function, is_public, get_public_methods, get_export_modules_by_dir,
+    get_module_trees_by_dir, get_doc_string, create_passthrough_args,
+    gen_sources, gen_formatted_sources, process_sources, parse_args, IMPORTS,
+    YAPF_STYLE
+)
+
+
+@pytest.fixture
+def pass_through_source():
+
+    source = '''def func():
+    """"""
+
+def one_arg_func(one):
+    """"""
+
+def two_args_func(one, two):
+    """"""
+
+def var_arg_func(one, *varargs):
+    """"""
+
+def kwonly_arg_func(one, *varargs, kwarg=''):
+    """"""
+
+def kwvar_arg_func(one, **varkwargs):
+    """"""
+
+def all_args_func(one, two, *args, kwargone='', kwargtwo=10, **kwargs):
+    """"""
+'''
+    return source
+
+
+@pytest.fixture
+def source():
+    """ Create a function and async function compiled code 
+    """
+    source = '''from _run import _public
+
+
+class Test:
+
+    @_public
+    def public_func(self):
+        """With doc string"""
+
+    @_public
+    async def public_async_func(self):
+        """"""
+'''
+    return source
+
+
+@pytest.fixture
+def non_pub_source():
+    """ Create a function and async function compiled code 
+    """
+    non_pub_source = '''def func():
+    """"""
+
+
+async def async_func():
+    """"""
+
+
+class Test:
+
+    def non_public_func(self):
+        """"""
+
+    def non_pub_func(self):
+        """"""
+'''
+    return non_pub_source
+
+
+@pytest.fixture
+def module(source):
+    """Compile the source into an ast module
+    """
+    return ast.parse(source)
+
+
+@pytest.fixture
+def pass_through_module(pass_through_source):
+    """Compile the pass through source into an ast module
+    """
+    return ast.parse(pass_through_source)
+
+
+@pytest.fixture
+def mod_path(tmp_path, source, non_pub_source):
+    mod_path = str(tmp_path.joinpath('public_module.py'))
+    with open(mod_path, 'w') as mod_file:
+        mod_file.write(source)
+    non_pub_mod_path = str(tmp_path.joinpath('non_public_module.py'))
+    with open(non_pub_mod_path, 'w') as mod_file:
+        mod_file.write(non_pub_source)
+    return tmp_path
+
+
+def test_is_function(module):
+    for node in ast.walk(module):
+        if isinstance(node, ast.FunctionDef):
+            assert is_function(node) is True
+        if isinstance(node, ast.AsyncFunctionDef):
+            assert is_function(node) is True
+    assert is_function(module) is False
+
+
+def test_is_public(module):
+    for node in ast.walk(module):
+        if is_function(node) and node.name == 'public_func':
+            assert is_public(node) is True
+        if is_function(node) and node.name == 'non_public_func':
+            assert is_public(node) is False
+
+
+@pytest.mark.skipif(
+    sys.version_info >= (3, 8), reason="requires on dev version"
```

Comment:
Sorry, I have not read the whole PR, but I don't understand this comment. I think the test only works on stable versions, so maybe the reason should read "requires a stable version" instead? ditto for the other occurrences below.

Reply:
Hi @pquentin, right thanks for pointing this out, clearly a typo. I am not sure, but maybe I should reference https://github.com/berkerpeksag/astor/pull/121

which will hopefully help to remove the skipif.

I'll fix this.